### DISCUSSION TITLE
Relationship between bbox and resolution

### DISCUSSION THREAD

**Discussion by trulymust** [pymupdf/PyMuPDF#1792] [Q&A] [answered]
Hello, developer, I want to ask about the relationship between the bbox and the DPI of the picture. For example, I generated the bbox in the default size of PDF, but the resolution was increased by 2 times when exporting the picture. How should I change the bbox? Or how should I generate the bbox in the resolution I want to export at the beginning?

**Comment by JorjMcKie** [comment]
You seem to talk about images created by pixmap generations.

There is no relationship between the two things. You have to actively set the dpi if another than the default is needed.
If you use the dpi parameter of `page.get_pixmap(dpi=xxx)`, it will be stored in the resulting PNG image - and should be respected as such by any software downstream.
If you instead use the matrix parameter, this automatism does not happen. Mainly because you can set reslution for x- and y-dimension to different values.
But you can always set dpi via `Pixmap.set_dpi(xres, yres)` which will land in the final image file too.

**Reply by trulymust** [reply]
thanks for your reply, I want to talk about how to change the bbox after the resolution is increased so that it can be displayed in the correct position after the resolution is increased

**Reply by JorjMcKie** [reply]
The location and size of the image on / in the medium you choose (a GUI window or a PDF page?) - this is the bbox you are talking about, right?
This bbox does not change whatever you do with the resolution.

**Reply by trulymust** [reply]
sorry, my english is poor, sorry to disturb you so long time.
![test_1](https://user-images.githubusercontent.com/55483183/176479906-48873dc4-dd0f-422f-ba00-dd247e6ccbd9.png)
this image is default size and the bbox is correctly.
![test_1](https://user-images.githubusercontent.com/55483183/176480097-2202637f-0aad-4de4-b35b-9a367e795abb.png)
But the resolution of this picture is quadrupled and the bbox is Original size.
I want to know how to modify bbox when the resolution is enlarged.

**Reply by JorjMcKie** [reply]
> sorry to disturb you so long time.

No worries - this is what these discussions are there for 😁!
Can you tell me exactly what you are trying to do? What am I seeing looking at your picture? A GUI, a PDF page?
What is your code to make the image and to put the image to the display device?

**Reply by trulymust** [reply]
Thanks, It's very kind of you. I have a batch of samples to mark the position of each line of text. However, the default resolution for saving PDF pages as images is too low to meet the requirements of my model. 
This is the code to save the picture：
```
        zoom_x = 1.0
        zoom_y = 1.0
        trans = fitz.Matrix(zoom_x, zoom_y).prerotate(rotate)
        pm = page.get_pixmap(matrix=trans, alpha=False)
        pm.save(imgName)
```
Bbox rendering is normal when the zoom factor is 1.0.
The picture above appears when the zoom factor is changed, Unable to correctly render the rectangular box of bbox. 
This is my rendering code：
```
def drawRectangle(bbox, image_path):
    image = Image.open(image_path)
    draw = ImageDraw.Draw(image)
    draw.rectangle([bbox[0], bbox[1], bbox[2], bbox[3]], outline='green', width=1
[code.zip](https://github.com/pymupdf/PyMuPDF/files/9012314/code.zip)
)
    del draw
    image.save(image_path)
```
So I want to know how to calculate the coordinates of bbox when the zoom factor increases.
Thank you.

**Comment by JorjMcKie** [comment, marked as answer]
In your code you have to actively set the dpi, otherwise it will not be made part of the PNG you are saving:
```python
zoom_x = 1.0
zoom_y = 1.0
trans = fitz.Matrix(zoom_x, zoom_y).prerotate(rotate)
pm = page.get_pixmap(matrix=trans, alpha=False)
pm.set_dpi(x_resolution, y_resolution)  # <=== !!!
pm.save(imgName)
```
So if I am computing correctly, for a zoom factor of 2 (which corresponds to 2 * 72 = 144 dpi), you should use `pm.set_dpi(144, 144)`.
The coordinates of the bbox should not be changed.

**Reply by trulymust** [reply]
OK, I'll try it. thank you. have a good day.

**Comment by JorjMcKie** [comment]
BTW you are aware that you do not need to save the pixmap as an image if all you want to do is using i directly afterwards?
Instead of `pixmap.save()` use `pixmap.tobytes()`. Or - you are obviously using PIL / Pillow - use `pixmap.pil_tobytes()` which will directly create a Pillow `Image`.
